apps/confidential/confidential_test.go

```diff
@@ -812,6 +812,119 @@ func TestWithTenantID(t *testing.T) {
 	}
 }
 
+func TestWithInstanceDiscovery(t *testing.T) {
+	accessToken := "*"
+	stackurl := "stack.local"
+	host := fmt.Sprintf("https://%s/", stackurl)
+	for _, tenant := range []string{
+		"adfs",
+		"98b8267d-e97f-426e-8b3f-7956511fd63f",
+	} {
+		for _, method := range []string{"authcode", "credential", "obo"} {
+			t.Run(method, func(t *testing.T) {
+				authority := host + tenant
+				cred, err := NewCredFromSecret("secret")
+				if err != nil {
+					t.Fatal(err)
+				}
+				idToken, refreshToken := "", ""
+				mockClient := mock.Client{}
+				if method == "obo" {
+					idToken = mock.GetIDToken(tenant, authority)
+					refreshToken = "refresh-token"
+				}
+				mockClient.AppendResponse(mock.WithBody(mock.GetTenantDiscoveryBody(stackurl, tenant)))
+				mockClient.AppendResponse(
+					mock.WithBody(mock.GetAccessTokenBody(accessToken, idToken, refreshToken, "", 3600)),
+				)
+				client, err := New("client-id", cred, WithAuthority(authority), WithHTTPClient(&mockClient), WithInstanceDiscovery(false))
+				if err != nil {
+					t.Fatal(err)
+				}
+				ctx := context.Background()
+				if _, err = client.AcquireTokenSilent(ctx, tokenScope); err == nil {
+					t.Fatal("silent auth should fail because the cache is empty")
+				}
+				var ar AuthResult
+				switch method {
+				case "authcode":
+					ar, err = client.AcquireTokenByAuthCode(ctx, "auth code", "https://localhost", tokenScope)
+				case "credential":
+					ar, err = client.AcquireTokenByCredential(ctx, tokenScope)
+				case "obo":
+					ar, err = client.AcquireTokenOnBehalfOf(ctx, "assertion", tokenScope)
+				default:
+					t.Fatal("test bug: no test for " + method)
+				}
+				if err != nil {
+					t.Fatal(err)
+				}
+				if ar.AccessToken != accessToken {
+					t.Fatalf(`unexpected access token "%s"`, ar.AccessToken)
+				}
+				if method == "obo" {
+					if ar, err = client.AcquireTokenOnBehalfOf(ctx, "assertion", tokenScope); err != nil {
+						t.Fatal(err)
+					}
+				} else if ar, err = client.AcquireTokenSilent(ctx, tokenScope); err != nil {
+					t.Fatal(err)
+				}
+				if ar.AccessToken != accessToken {
+					t.Fatal("cached access token should match the one returned by AcquireToken...")
+				}
+			})
+		}
+	}
+}
+
+func TestWithPortAuthority(t *testing.T) {
+	accessToken := "*"
+	lmo := "stack.local"
+	port := ":3001"
+	host := lmo + port
+	tenant := "00000000-0000-0000-0000-000000000000"
+	authority := fmt.Sprintf("https://%s%s/%s", lmo, port, tenant)
+	cred, err := NewCredFromSecret("secret")
+	if err != nil {
+		t.Fatal(err)
+	}
+	idToken, refreshToken, URL := "", "", ""
+	mockClient := mock.Client{}
+	//2 calls to instance discovery are made because Host is not trusted
+	mockClient.AppendResponse(mock.WithBody(mock.GetInstanceDiscoveryBody(host, tenant)))
+	mockClient.AppendResponse(mock.WithBody(mock.GetInstanceDiscoveryBody(host, tenant)))
```

Comment:
One request is sent out during the silentToken request, and the other is sent during the AuthCode request. Since the host isn't trusted, two requests are sent.

Reply:
I opened #351 for this kind of thing